File: backend_v2/app/core/roles.py
"""
SSOT (Single Source of Truth) de constantes de roles del sistema.

IMPORTANTE: Este módulo documenta los roles seed del sistema.
NO se usa para autorización en runtime. La autorización real es siempre
vía ServicioAuth.tiene_acceso_panel_admin() o equivalente, que consulta
dinámicamente PermisoRol JOIN ModuloSistema WHERE categoria='panel'.

Si en el futuro se añade un nuevo rol:
1. Añadirlo a ROLES_SEED aquí.
2. Registrarlo en admin_router.py:271-280 (RolSistema seed).
3. Asignarle permisos de categoria='panel' desde la UI de admin.
4. El endpoint /verify-admin lo aceptará automáticamente.

Mantener esta constante sincronizada con frontend/src/constants/roles.ts
(espejo frontend). El test test_roles_ssot_syncronizado valida que ambas
listas no diverjan.
"""
from typing import FrozenSet

ROLES_SEED: FrozenSet[str] = frozenset({
    "admin",
    "admin_sistemas",
    "admin_mejoramiento",
    "director",
    "analyst",
    "manager",
    "viaticante",
    "usuario",
})


ROLES_ADMIN_PANEL: FrozenSet[str] = frozenset({
    "admin",
    "admin_sistemas",
    "admin_mejoramiento",
    "director",
    "analyst",
})


def es_rol_valido(rol: str) -> bool:
    """Verifica si un rol existe en el sistema."""
    return rol in ROLES_SEED

File: frontend/src/constants/roles.ts
/**
 * SSOT (Single Source of Truth) espejo frontend de `backend_v2/app/core/roles.py`.
 *
 * IMPORTANTE: Mantener sincronizado con el backend. Si añades un rol
 * nuevo, actualiza AMBOS archivos y el test test_roles_ssot_syncronizado
 * en el backend.
 *
 * Esta constante es DOCUMENTATIVA: la autorización real siempre se valida
 * en el backend vía ServicioAuth.tiene_acceso_panel_admin(). Esta lista
 * se usa SOLO para el guard de UI (mostrar/ocultar elementos interactivos).
 */

export const ROLES_SEED = [
  'admin',
  'admin_sistemas',
  'admin_mejoramiento',
  'director',
  'analyst',
  'manager',
  'viaticante',
  'usuario',
] as const;

export const ROLES_ADMIN_PANEL = [
  'admin',
  'admin_sistemas',
  'admin_mejoramiento',
  'director',
  'analyst',
] as const;

export type RolSeed = (typeof ROLES_SEED)[number];
export type RolAdminPanel = (typeof ROLES_ADMIN_PANEL)[number];

export function esRolValido(rol: string): rol is RolSeed {
  return (ROLES_SEED as readonly string[]).includes(rol);
}
